src/storage.ts

```diff
@@ -550,6 +626,132 @@ export class Storage extends Service {
+   * @property {HmacKey} hmacKey The HmacKey object.
+   * @property {string} secret The HMAC key secret used to access XML API.
+   */
+  /**
+   * @typedef {array} CreateHmacKeyResponse
+   * @property {HmacKeyResource} 0 The HMAC key resource.
+   */
+  /**
+   * @callback CreateHmacKeyCallback Callback function.
+   * @param {?Error} err Request error, if any.
+   * @param {HmacKeyResource} resource An object containing the HMAC key's secret and the HmacKey object.
+   * @param {object} apiResponse The raw API response.
+   */
+  /**
+   * Create an HMAC key associated with an service account to authenticate
+   * requests to the Cloud Storage XML API.
+   *
+   * @see [HMAC keys documentation]{@linkhttps://cloud.google.com/storage/docs/authentication/hmackeys}
+   *
+   * @param {string} serviceAccountEmail The service account's email address
+   *     with which the HMAC key is created for.
+   * @param {CreateHmacKeyCallback} [callback] Callback function.
+   * @return {Promise<CreateHmacKeyResponse>}
+   *
+   * @example
+   * const {Storage} = require('google-cloud/storage');
+   * const storage = new Storage();
+   *
+   * // Replace with your service account's email address
+   * const serviceAccountEmail =
+   *   'my-service-account@appspot.gserviceaccount.com';
+   *
+   * storage.createHmacKey(serviceAccountEmail, function(err, hmacKeyResource) {
+   *   if (!err) {
+   *     const secret = hmacKey.secret;
+   *     // Securely store the secret for use with the XML API.
+   *   }
+   * });
+   *
+   * //-
+   * // If the callback is omitted, we'll return a Promise.
+   * //-
+   * storage.createHmacKey(serviceAccountEmail)
+   *   .then((response) => {
+   *     const [hmacKeyResource] = response;
+   *     const secret = hmacKeyResource.secret;
+   *     // Securely store the secret for use with the XML API.
+   *   });
+   */
+  createHmacKey(
+    serviceAccountEmail: string,
+    optionsOrCb?: CreateHmacKeyOptions | CreateHmacKeyCallback,
+    cb?: CreateHmacKeyCallback
+  ): Promise<CreateHmacKeyResponse> | void {
+    if (typeof serviceAccountEmail !== 'string') {
+      throw new Error(
+        'The first argument must be a service account email to create an HMAC key.'
+      );
+    }
+
+    const {options, callback} = normalize<
+      CreateHmacKeyOptions,
+      CreateHmacKeyCallback
+    >(optionsOrCb, cb);
+    const query = Object.assign({}, options, {serviceAccountEmail});
+
+    this.request(
+      {
+        method: 'POST',
+        uri: `/projects/${this.projectId}/hmacKeys`,
+        qs: query,
+      },
+      (err, resp: HmacKeyResourceResponse) => {
+        if (err) {
+          callback!(err, null, resp);
+          return;
+        }
+
+        const hmacKey = new HmacKey(this, resp.metadata.accessId);
+        hmacKey.metadata = resp.metadata;
+
+        const resource = {
+          hmacKey,
+          secret: resp.secret,
+        };
+
+        callback!(null, resource, resp);
```

Comment:
This signature would be the first to break from the convention across the libraries, where we strictly return the created resource instance. Can the secret be stored on the HmacKey instance?

Reply:
Unlike buckets or files where the API response is the resource itself, creating HMAC key returns a secret in addition to the resource. This makes it clearer that the secret is only returned on creation.

Reply:
Could it be `callback(null, hmacKey, resp.secret, resp)`? Or maybe if we said in the docs where `secret` can be found within the `resp` argument? I think the consistent hierarchy is important to maintain, where `createThing` returns `Thing`, unboxed.